src/Avalonia.Visuals/CornerRadius.cs

```diff
@@ -26,49 +28,127 @@ public CornerRadius(double uniformRadius)
         public CornerRadius(double top, double bottom)
         {
-            TopLeft = TopRight = top;
+            TopLeft    = TopRight    = top;
             BottomLeft = BottomRight = bottom;
+
+            TopLeftRadiusX     = TopLeftRadiusY     = TopLeft;
+            TopRightRadiusX    = TopRightRadiusY    = TopRight;
+            BottomRightRadiusX = BottomRightRadiusY = BottomRight;
+            BottomLeftRadiusX  = BottomLeftRadiusY  = BottomLeft;
+
+            _isCircular = true;
         }
-        public CornerRadius(double topLeft, double topRight, double bottomRight, double bottomLeft)
+
+        public CornerRadius(
+            double topLeft,
+            double topRight,
+            double bottomRight,
+            double bottomLeft)
         {
-            TopLeft = topLeft;
-            TopRight = topRight;
+            TopLeft     = topLeft;
+            TopRight    = topRight;
             BottomRight = bottomRight;
-            BottomLeft = bottomLeft;
+            BottomLeft  = bottomLeft;
+
+            TopLeftRadiusX     = TopLeftRadiusY     = TopLeft;
+            TopRightRadiusX    = TopRightRadiusY    = TopRight;
+            BottomRightRadiusX = BottomRightRadiusY = BottomRight;
+            BottomLeftRadiusX  = BottomLeftRadiusY  = BottomLeft;
+
+            _isCircular = true;
+        }
+
+        public CornerRadius(
+            double topLeftRadiusX,
+            double topLeftRadiusY,
+            double topRightRadiusX,
+            double topRightRadiusY,
+            double bottomRightRadiusX,
+            double bottomRightRadiusY,
+            double bottomLeftRadiusX,
+            double bottomLeftRadiusY)
+        {
+            TopLeftRadiusX     = topLeftRadiusX;
+            TopLeftRadiusY     = topLeftRadiusY;
+            TopRightRadiusX    = topRightRadiusX;
+            TopRightRadiusY    = topRightRadiusY;
+            BottomRightRadiusX = bottomRightRadiusX;
+            BottomRightRadiusY = bottomRightRadiusY;
+            BottomLeftRadiusX  = bottomLeftRadiusX;
+            BottomLeftRadiusY  = bottomLeftRadiusY;
+
+            TopLeft     = (TopLeftRadiusX     + TopLeftRadiusY)     / 2.0;
+            TopRight    = (TopRightRadiusX    + TopRightRadiusY)    / 2.0;
+            BottomRight = (BottomRightRadiusX + BottomRightRadiusY) / 2.0;
+            BottomLeft  = (BottomLeftRadiusX  + BottomLeftRadiusY)  / 2.0;
+
+            _isCircular = (TopLeftRadiusX     == TopLeftRadiusY &&
+                           TopRightRadiusX    == TopRightRadiusY &&
+                           BottomRightRadiusX == BottomRightRadiusY &&
+                           BottomLeftRadiusX  == BottomLeftRadiusY);
         }
 
         /// <summary>
-        /// Radius of the top left corner.
+        /// Gets the circular radius of the top left corner.
         /// </summary>
         public double TopLeft { get; }
 
+        public double TopLeftRadiusX { get; }
+        public double TopLeftRadiusY { get; }
+
```

Comment:
These could also be combined with a `Size` or `Point` struct.

Reply:
This would also allow a new constructor taking a single Size/Point to set uniform elliptical corner radius. The different type allows it to be differentiated from a uniform circular corner radius which takes a single double value. The same can be done for other constructors.

Reply:
Changed to `Size` everywhere. It seems to be the cleanest API after consideration of individual `double` X/Y properties or the `Point` struct.

Reply:
Wouldn't a new Struct EllipticalRadius with RadiusX and RadiusY make sense here? You can then have computed properties for IsCircular and AvarageRadius for example.

Reply:
I agree, then perhaps we can make them convertible to and from Size

Reply:
Yes, I also realized when I switched to Size that CornerRadius itself is kind of wrong. It should be CornerRadii and be composed of individual CornerRadius structs. I like your name idea though and the code is evolving real-time just to see if all this even makes sense.

I guess I didn't really want to create a new struct though property naming issues aren't solved yet. RadiusX/Y are more traditional names and make more sense so its likely justified. Width/Height didn't completely work but it made more sense to use Size rather than Point.

Anyway, I think you are correct and leaving CornerRadius as-is (instead of renaming to CornerRadii) and adding a new EllipticalRadius is the best overall solution.

Reply:
> You can then have computed properties for IsCircular and AvarageRadius

I could have done computed properties now for IsCircular but decided against at that code runs every time its called. Seems more performant to pre-compute since this struct is immutable. Also AverageRadius can never be pre-computed as it is the existing TopLeft/etc. properties.

Reply:
@timunie Sorry, I didn't fully understand what you were saying at first. I've made the updates though -- please take a look.

Reply:
Hey @robloo , 

my idea was that: 

```c#
        public double TopLeft { get; }

        public double TopLeftRadiusX { get; }
        public double TopLeftRadiusY { get; }

```

will be replaced by 

```c# 
      public MyCornerRadiusStruct TopLeft {get; } 
```

which will expose all needed properties and Methods. But I am also happy with the idea you suggested. 

Happy coding
Tim

Reply:
@timunie Ok, I see. That would probably be the correct approach if this was a brand-new implementation (along with renaming CornerRadius itself to CornerRadii). However, I'm trying to keep full compatibility with the current implementation in Avalonia and the implementation in UWP. Therefore, I need to keep the `double TopLeft` properties for circular radius separately. These are pre-existing and I don't want to break all that code.

Reply:
I fully agree. Just wanted to bring this idea while many breaking changes are ongoing, so the core team can still decide to use this implemention.

Otherwise it looks good to me. Looking forward 😃

Happy coding and a great weekend 👍
Tim